### ISSUE TITLE
Switch APIs from using discrete optional arguments to using an options argument where it makes sense

### ISSUE THREAD

**Issue by hossam-nasr** [Azure/azure-functions-durable-js#441]
Based on discussion [here](https://github.com/Azure/azure-functions-durable-js/pull/435#discussion_r1088287211). This issue is a generalization of #415 and #435, to switch all of our APIs from using optional discrete arguments to accepting an options object argument, where applicable and where it makes sense.

**Comment by hossam-nasr** [comment]
APIs affected/proposed:

**In DurableOrchestrationClient:**

<table>
<tr>
<th> Before </th>
<th> Proposed change </th>
</tr>
<tr>
<td>

```TS
getStatus(
    instanceId: string,
    showHistory?: boolean,
    showHistoryOutput?: boolean,
    showInput?: boolean
): Promise<DurableOrchestrationStatus>
```
</td>
<td>

```TS
getStatus(
    instanceId: string, 
    options?: GetStatusOptions
): Promise<DurableOrchestrationStatus>
```
</td>
</tr>
<tr>
<td>

```TS
getStatusBy(
    createdTimeFrom: Date | undefined,
    createdTimeTo: Date | undefined,
    runtimeStatus: OrchestrationRuntimeStatus[]
): Promise<DurableOrchestrationStatus[]>
```

</td>
<td>

```TS
getStatusBy(
    options: SelectionOptions
): Promise<DurableOrchestrationStatus[]>
```

</td>
</tr>
<tr>
<td>

```TS
purgeInstanceHistoryBy(
    createdTimeFrom: Date,
    createdTimeTo?: Date,
    runtimeStatus?: OrchestrationRuntimeStatus[]
): Promise<PurgeHistoryResult>
```

</td>
<td>

```TS
purgeInstanceHistoryBy(
    options: SelectionOptions
): Promise<PurgeHistoryResult>
```

</td>
</tr>
<tr>
<td>

```TS
raiseEvent(
    instanceId: string,
    eventName: string,
    eventData: unknown,
    taskHubName?: string,
    connectionName?: string
): Promise<void>
```

</td>
<td>

```TS
raiseEvent(options: RaiseEventOptions): Promise<void>
```

</td>
</tr>
<tr>
<td>

```TS
readEntityState<T>(
    entityId: EntityId,
    taskHubName?: string,
    connectionName?: string
): Promise<EntityStateResponse<T>>
```

</td>
<td>

```TS
readEntityState<T>(
    entityId: EntityId,
    options?: TaskHubOptions
): Promise<EntityStateResponse<T>>
```

</td>
</tr>
<tr>
<td>

```TS
rewind(
    instanceId: string,
    reason: string,
    taskHubName?: string,
    connectionName?: string
): Promise<void>`
```

</td>
<td>

```TS
rewind(
    instanceId: string, 
    reason: string, 
    options?: TaskHubOptions
): Promise<void>
```

</td>
</tr>
<tr>
<td>

```TS
signalEntity(
    entityId: EntityId,
    operationName?: string,
    operationContent?: unknown,
    taskHubName?: string,
    connectionName?: string
): Promise<void>
```
</td>
<td>

```TS
signalEntity(
    entityId: EntityId, 
    options?: SignalEntityOptions
): Promise<void>
```
</td>
</tr>
<tr>
<td>

```TS
startNew(
    orchestratorFunctionName: string,
    instanceId?: string,
    input?: unknown
): Promise<string>
```

</td>
<td>

```TS
// this change is already merged
startNew(
    orchestratorFunctionName: string, 
    options?: StartNewOptions
): Promise<string>
```

</td>
</tr>
</table>

**In DurableOrchestrationContext:**

<table>
<tr>
<th> Before </th>
<th> Proposed Change </th>
</tr>

<tr>
<td>

```TS
callSubOrchestrator(
    name: string,
     input?: unknown, 
    instanceId?: string
): Task
```

</td>
<td>

```TS
// This is effectively equivalent
// to the startNew() change
callSubOrchestrator(
    name: string, 
    options?: StartNewOptions
): Task
```

</td>
</tr>
<tr>
<td>

```TS
callSubOrchestratorWithRetry(
    name: string,
    retryOptions: RetryOptions,
    input?: unknown,
    instanceId?: string
): Task
```

</td>
<td>

```TS
callSubOrchestratorWithRetry(
    name: string,
    retryOptions: RetryOptions,
    options?: CallSubOrchestratorOptions
): Task
```

</td>
</tr>
<tr>
<td>

```TS
callHttp(
    method: string,
    uri: string,
    content?: string | object,
    headers?: { [key: string]: string },
    tokenSource?: TokenSource,
    asynchronousPatternEnabled = true
): Task
```

</td>
<td>

```TS
// This change is already merged
callHttp(options: CallHttpOptions): Task
```

</td>
</tr>
</table>

**Comment by davidmrdavid** [comment]
Thanks @hossam-nasr. 
I like the idea of modernizing these through an options object, but given the size of the list (over 10 APIs), I would have to request that any change to these should be **backwards compatible.**

Have we already made `callHttp` and `startNew` backwards compatible with the old style?

**Comment by ejizba** [comment]
Unfortunately we have to do overloads to be backwards compatible, and overloads pretty negatively impact intellisense and build errors. I would prefer we choose between a clean yes or no on this change

**Comment by davidmrdavid** [comment]
Understandable. At this time, I do not feel comfortable taking the _entirety_ of this list as breaking changes.
Let me reach out internally for some ideas on how to move forward.

**Comment by hossam-nasr** [comment]
Closed by #482
